File: client/src/core/base/SceneFactory.java
package core.base;

import core.input.InputController;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.HashMap;

public class SceneFactory {
	protected final Logger logger = LogManager.getLogger();

	private Window window;
	private SceneController controller;

	private InputController input;

	public SceneFactory(SceneController controller, Window window, InputController input) {
		this.controller = controller;
		this.window = window;
		this.input = input;
	}

	public Scene createScene(Class<? extends Scene> sceneClass, Object... args) {
		HashMap<Class<?>, Object> mapArgs = mapArguments(args);
		// TODO: delegate to scene classes
		logger.error("Can't create scene of type {}", sceneClass);
		return null;
	}

	private HashMap<Class<?>, Object> mapArguments(Object[] args) {
		HashMap<Class<?>, Object> result = new HashMap<>();
		for (Object arg : args)
			result.put(arg.getClass(), arg);
		return result;
	}
}

File: client/src/core/base/SceneFactory.kt
package core.base

import org.apache.logging.log4j.LogManager
import java.util.*
import java.util.function.Supplier

class SceneFactory(
) {

	protected val logger = LogManager.getLogger(javaClass)!!
	private val generators: HashMap<Class<out Scene>, Supplier<Scene>> = HashMap()

	fun createScene(sceneClass: Class<out Scene>): Scene? {
		return generators[sceneClass]?.get()
	}

	fun <T : Scene> register(sceneClass: Class<T>, generator: Supplier<Scene>) {
		generators[sceneClass] = generator
	}

	private fun mapArguments(args: Array<Any>): HashMap<Class<*>, Any> {
		val result = HashMap<Class<*>, Any>()
		for (arg in args) result[arg.javaClass] = arg
		return result
	}

}
